ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2EventStrategy.scala

```diff
@@ -45,15 +45,21 @@ object H2EventStrategy extends EventStrategy {
         cSQL"(${H2QueryStrategy.arrayIntersectionNonEmptyClause(witnessesColumnName, wildCardParties, stringInterning)})" :: Nil
     }
     val partiesTemplatesClauses =
-      filterParams.partiesAndTemplates.iterator.map { case (parties, templateIds) =>
+      filterParams.partiesAndTemplates.iterator.flatMap { case (parties, templateIds) =>
         val clause =
           H2QueryStrategy.arrayIntersectionNonEmptyClause(
             witnessesColumnName,
             parties,
             stringInterning,
           )
-        val templateIdsArray = templateIds.view.map(_.toString).toArray
-        cSQL"( ($clause) AND (template_id = ANY($templateIdsArray)) )"
+        val templateIdsArray: Array[java.lang.Integer] =
+          templateIds.view
+            .map(stringInterning.templateId.tryInternalize)
+            .flatMap(_.toList)
+            .map(Int.box)
+            .toArray
+        if (templateIdsArray.isEmpty) Iterator.empty
```

Comment:
```suggestion
        if (templateIdsArray.isEmpty) Iterator(cSQL"FALSE")
```
This would be an alternative (I think) to matching for an empty list below.